storage/migrations/0009_oasis_3_emerald_init.up.sql

```diff
@@ -16,10 +19,56 @@ CREATE TABLE IF NOT EXISTS oasis_3.emerald_rounds
   messages_hash    TEXT NOT NULL,
   in_messages_hash TEXT NOT NULL,
 
+  -- TODO: These should also be NOT NULL, but they're populated separately.
+  num_transactions INTEGER,
+  gas_used         BIGINT,
+  size             INTEGER,
+
   -- Arbitrary additional data.
   extra_data JSON
 );
 
+CREATE TABLE IF NOT EXISTS oasis_3.emerald_transactions
+(
+  round BIGINT NOT NULL,
+  tx_index INTEGER NOT NULL,
+  tx_hash TEXT NOT NULL,
+  tx_eth_hash TEXT,
+  PRIMARY KEY (round, tx_index)
+);
+
+CREATE INDEX ix_emerald_transactions_tx_hash ON oasis_3.emerald_transactions(tx_hash);
+CREATE INDEX ix_emerald_transactions_tx_eth_hash ON oasis_3.emerald_transactions(tx_eth_hash);
+
+CREATE TABLE IF NOT EXISTS oasis_3.emerald_transaction_signers
+(
+  round BIGINT NOT NULL,
+  tx_index INTEGER NOT NULL,
+  signer_index INTEGER NOT NULL,
+  signer_address TEXT NOT NULL,
+  nonce BIGINT NOT NULL,
+  PRIMARY KEY (round, tx_index, signer_index)
+);
+
+CREATE INDEX ix_emerald_transaction_signers_signer_address_signer_nonce ON oasis_3.emerald_transaction_signers(signer_address, nonce);
+
+CREATE TABLE IF NOT EXISTS oasis_3.emerald_related_transactions
+(
+  account_address TEXT NOT NULL,
+  tx_round BIGINT NOT NULL,
+  tx_index INTEGER NOT NULL
+);
+
+CREATE INDEX ix_emerald_related_transactions_address_height_index ON oasis_3.emerald_related_transactions(account_address, tx_round, tx_index);
+
+CREATE TABLE IF NOT EXISTS oasis_1.address_preimages
+(
+  address TEXT NOT NULL PRIMARY KEY,
+  context_identifier TEXT NOT NULL,
```

Comment:
```
CREATE TYPE address_context_identifier AS ENUM ('oasis-runtime-sdk/address: secp256k1eth', 'oasis-runtime-sdk/address: sr25519', 'oasis-runtime-sdk/address: multisig', 'oasis-runtime-sdk/address: module');
```
then
```suggestion
  eth_address_context address_context_identifier NOT NULL,
```

I'm not 100% if this context is really for the eth address or the oasis address :/

In Emerald, when can this be `!= 'oasis-runtime-sdk/address: secp256k1eth'`?

Reply:
this table is not specific to emerald

> if this context is really for the eth address or the oasis address

- an eth address exists independently of the fact that some project called oasis invented a way to derive an oasis address from it
- every oasis address is derived from a context (id, version) and a piece of data. unless I'm forgetting about some exceptions

Reply:
"this table is not specific to emerald" -> I missed this bit. Thank you.

The rest of the comment stands though:
- We should introduce the `address_context_identifier` enum, both to be self-documenting and to prevent weird data from creeping in
- I would much prefer `TEXT` over `BYTEA`, because inputs, outputs, and debugging will always be text. Paratime-specific tables with which this one will be `JOIN`-ed will also have `TEXT` in their paratime-specific address columns. So having `TEXT` in this column will make it much easier to develop. If I understand the columns correctly (still slightly dubious) it's possible that two paratimes might use the same context and thus `address_data`, but disagree on the canonical stringification. I'm fine with that -- sounds unlikely in practice, and if it does happen, we can worry about it later, e.g. by introducing conversion functions (which we'd need with `BYTEA` anyway) or additional precomputed columns.

Reply:
> prevent weird data from creeping in

I'm not convinced about that threat. but it's good for another reason, it seems like a nice way to represent the highly repetitive values compactly short of making a separate table of known contexts.

however, it'll be annoying to keep that enum declaration in sync if we have to add contexts. as I understand it, we'll need a db migration

> TEXT

ya let's do that. lots of other places where we store binary data in various string representations

> debugging will always be text

db tools seem to display bytea values as text already. and in go we'll quickly convert to byte slice, so there'd be no difference there

> Paratime-specific tables with which this one will be JOIN-ed will also have TEXT in their paratime-specific address columns

shouldn't be joining on this table in that direction. if we have the data and want to get the derived address, we should run the hash function normally. joining by the `address` column is fine, and that column uses the predominant bech32 encoded format.

> ... address_data, but disagree on the canonical stringification

let's be careful about it :shrug: . let's use uh... hex. is that what most of the db uses to store opaque binary data?

> assumes that address_data is the paratime address

in some derivations it is an address, but this table is general, and all we know is that it's the data. could also be a CBOR-encoded multisig config, for example.

> should (and can?) we store the derived addresses

that not being the case,

Reply:
more on weird data creeping in: the oasis-core codebase contains a registry of address derivation contexts https://github.com/oasisprotocol/oasis-core/blob/v22.1.11/go/common/crypto/address/context.go . using a database enum would create a situation where we have to keep two copies of that registry in sync.

maybe what we could do is use `address.Context` in the AddressPreimageData struct so that it's clearer that we don't make up any unregistered context IDs or versions